src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -111,163 +460,135 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
-      result = r->allocate(size, req.type());
-      assert (result != nullptr, "Allocation must succeed: free " SIZE_FORMAT ", actual " SIZE_FORMAT, free, size);
+    if (adjusted_size >= req.min_size()) {
+      result = r->allocate(adjusted_size, req.type());
+      log_debug(gc, free)("Allocated " SIZE_FORMAT " words (adjusted from " SIZE_FORMAT ") for %s @" PTR_FORMAT
+                          " from %s region " SIZE_FORMAT ", free bytes remaining: " SIZE_FORMAT,
+                          adjusted_size, req.size(), ShenandoahAllocRequest::alloc_type_to_string(req.type()), p2i(result),
+                          free_memory_type_name(_free_sets.membership(r->index())),  r->index(), r->free());
+      assert (result != nullptr, "Allocation must succeed: free " SIZE_FORMAT ", actual " SIZE_FORMAT, free, adjusted_size);
+      req.set_actual_size(adjusted_size);
+    } else {
+      log_trace(gc, free)("Failed to shrink TLAB or GCLAB request (" SIZE_FORMAT ") in region " SIZE_FORMAT " to " SIZE_FORMAT
+                          " because min_size() is " SIZE_FORMAT, req.size(), r->index(), adjusted_size, req.min_size());
     }
   } else {
+    size_t size = req.size();
     result = r->allocate(size, req.type());
+    if (result != nullptr) {
+      // Record actual allocation size
+      log_debug(gc, free)("Allocated " SIZE_FORMAT " words for %s @" PTR_FORMAT
+                          " from %s region " SIZE_FORMAT ", free bytes remaining: " SIZE_FORMAT,
+                          size, ShenandoahAllocRequest::alloc_type_to_string(req.type()), p2i(result),
+                          free_memory_type_name(_free_sets.membership(r->index())),  r->index(), r->free());
+      req.set_actual_size(size);
+    }
   }
 
   if (result != nullptr) {
     // Allocation successful, bump stats:
     if (req.is_mutator_alloc()) {
-      increase_used(size * HeapWordSize);
-    }
-
-    // Record actual allocation size
-    req.set_actual_size(size);
+      _free_sets.increase_used(Mutator, req.actual_size() * HeapWordSize);
+    } else {
+      assert(req.is_gc_alloc(), "Should be gc_alloc since req wasn't mutator alloc");
 
-    if (req.is_gc_alloc()) {
+      // For GC allocations, we advance update_watermark because the objects relocated into this memory during
+      // evacuation are not updated during evacuation.
       r->set_update_watermark(r->top());
     }
   }
 
-  if (result == nullptr || has_no_alloc_capacity(r)) {
-    // Region cannot afford this or future allocations. Retire it.
-    //
-    // While this seems a bit harsh, especially in the case when this large allocation does not
-    // fit, but the next small one would, we are risking to inflate scan times when lots of
-    // almost-full regions precede the fully-empty region where we want allocate the entire TLAB.
-    // TODO: Record first fully-empty region, and use that for large allocations
+  if (alloc_capacity(r) < PLAB::min_size() * HeapWordSize) {
```

Comment:
Why not `TLAB::min_size`?

Reply:
Good question.  You've caused me to dig a little to respond:

size_t PLAB::min_size() {
  // Make sure that we return something that is larger than AlignmentReserve                                                                                 
  return align_object_size(MAX2(MinTLABSize / HeapWordSize, (size_t)oopDesc::header_size())) + CollectedHeap::lab_alignment_reserve();
}

There is no TLAB::min_size() as far as I can tell.  There is MinTLABSize (default 2K unless redefined on command line), but that value appears to get rounded up by alignment in the definition of PLAB::min_size().  I think PLAB:min_size() is what we want here (and elsewhere within this source file).

What do you think?

Reply:
Interesting. Yeah, that looks like what we want.